auth/default_authorizer.cc

```diff
@@ -340,7 +340,20 @@ const resource_set& default_authorizer::protected_resources() const {
 }
 
 bool default_authorizer::safe(const cql3::statements::schema_altering_statement& stmt) const {
-    return !protected_resources().contains(make_data_resource(stmt.keyspace(), stmt.column_family()));
+    if (!protected_resources().contains(make_data_resource(stmt.keyspace(), stmt.column_family()))) {
+        return true;
+    }
+    if (auto alter = dynamic_cast<const cql3::statements::alter_table_statement*>(&stmt)) {
+        using alter_type = cql3::statements::alter_table_statement::type;
+        switch (alter->get_type()) {
+        case alter_type::add:
+        case alter_type::opts:
+            return true;
+        default:
+            return false;
+        }
+    }
+    return false;
```

Comment:
Ok, so I suggest adding a type `alteration_properties` exposed by authorizer and imported by cql3 (which already knows about authorizer) where alter_statement descripts what it is going to do, and authorizer  examines. I don't think it should mirror everything alter_table_statement can do, just the bits that authorizer needs, and we can expand it on demand (hopefully there won't be a huge flood of things to protect).

Reply:
Let me try.  I'm not sure I'll like having a schema_altering_statement doppelgänger just for auth to consume; these things are expensive to maintain over time and confusing when out-of-sync.  Also, the best way for authorizer/authenticator/role_manager to express what they (dis)allow is via executable code, rather than a registered data structure.

Reply:
The idea is for alter_table_statement/drop_stable_statement to express what they are changing, rather than for the authorizer to express what it allows. The authorizer is still code, and alter_table_statement is already data driven.

Note the intermediary does not have to be as expressive as alter_table_statement.